File: Farmer/flutter_app/lib/models/sample.dart
// Sample collection and laboratory referral model

enum SampleStatus {
  recommended,
  collected,
  sentToLab,
  resultAvailable,
}

extension SampleStatusExt on SampleStatus {
  String get displayName {
    switch (this) {
      case SampleStatus.recommended:
        return 'Recommended';
      case SampleStatus.collected:
        return 'Collected';
      case SampleStatus.sentToLab:
        return 'Sent to Lab';
      case SampleStatus.resultAvailable:
        return 'Result Available';
    }
  }
}

class Sample {
  final String sampleId;
  final String caseId;
  final String animalId;
  final String animalTag;
  final String sampleType; // Blood, Faecal, Nasal swab, Tissue, etc.
  final String collectionDate;
  final String collectionLocation;
  final String reason;
  final String? laboratory;
  SampleStatus status;
  String? resultSummary;
  String? resultDate;
  final String collectedBy; // Vet name
  final DateTime createdAt;

  Sample({
    required this.sampleId,
    required this.caseId,
    required this.animalId,
    required this.animalTag,
    required this.sampleType,
    required this.collectionDate,
    required this.collectionLocation,
    required this.reason,
    this.laboratory,
    this.status = SampleStatus.recommended,
    this.resultSummary,
    this.resultDate,
    required this.collectedBy,
    DateTime? createdAt,
  }) : createdAt = createdAt ?? DateTime.now();

  Map<String, dynamic> toJson() => {
        'sampleId': sampleId,
        'caseId': caseId,
        'animalId': animalId,
        'sampleType': sampleType,
        'status': status.name,
        'laboratory': laboratory,
        'collectedBy': collectedBy,
        'createdAt': createdAt.toIso8601String(),
      };
}

File: backend/app/models/sample.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class SampleCreateRequest(BaseModel):
    case_id: str
    animal_id: Optional[str] = None
    animal_tag: str
    sample_type: str  # Blood, Faecal, Nasal Swab, Tissue, Milk, Urine, Other
    laboratory: str
    reason: str
    collection_location: Optional[str] = None


class SampleResultUpdateRequest(BaseModel):
    status: str  # RESULT_AVAILABLE, REJECTED, etc.
    result_summary: str
    result_date: Optional[str] = None


class SampleResponse(BaseModel):
    id: str
    sample_code: str
    case_id: str
    animal_id: Optional[str] = None
    animal_tag: str
    sample_type: str
    collection_date: str
    collection_location: Optional[str] = None
    reason: str
    laboratory: str
    status: str
    result_summary: Optional[str] = None
    result_date: Optional[str] = None
    collected_by: str
    created_at: datetime
